src/components/UI/EditableValueLabel.tsx

```diff
@@ -0,0 +1,199 @@
+import styled from '@emotion/styled';
+import PropTypes from 'prop-types';
+import React, {
+  KeyboardEventHandler,
+  Reducer,
+  useReducer,
+  useRef,
+  useState,
+} from 'react';
+import Overlay from 'react-bootstrap/lib/Overlay';
+import Tooltip from 'react-bootstrap/lib/Tooltip';
+import Button from 'UI/Button';
+import { InputElement } from 'UI/Input';
+import { IValidation, IValidationFunction } from 'utils/labelUtils';
+
+interface IEditableValueLabel {
+  allowEdit: boolean;
+  isNew: boolean;
+  onCancel(): void;
+  onEdit(): void;
+  onSave(value: string): void;
+  validationFunc: IValidationFunction;
+  value: string;
+}
+
+interface IReducerState extends IValidation {
+  internalValue: string;
+}
+
+const EditModeWrapper = styled.div`
+  display: flex;
+  align-items: center;
+  margin: 10px 0 10px 5px;
+`;
+
+const ValueInput = styled(InputElement)`
+  margin-right: 10px;
+`;
+
+const EditValueTooltip = styled(Tooltip)`
+  &.in {
+    opacity: 1;
+  }
+  .tooltip-inner {
+    background-color: ${({ theme }) => theme.colors.darkBlueLighter3};
+  }
+  &.top .tooltip-arrow {
+    border-top-color: ${({ theme }) => theme.colors.darkBlueLighter3};
+  }
+`;
+
+const ValidationErrorTooltip = styled(Tooltip)`
+  &.in {
+    opacity: 1;
+  }
+  .tooltip-inner {
+    background-color: ${({ theme }) => theme.colors.error};
+    color: black;
+  }
+  &.top .tooltip-arrow {
+    border-top-color: ${({ theme }) => theme.colors.error};
+  }
+`;
+
+const Editable = styled.span<{ allowEdit: boolean }>`
+  text-decoration: ${({ allowEdit }) => (allowEdit ? 'underline' : 'none')};
+  text-decoration-style: dotted;
+  cursor: ${({ allowEdit }) => (allowEdit ? 'pointer' : 'default')};
+  opacity: ${({ allowEdit }) => (allowEdit ? '1' : '0.6')};
+`;
+
+const EditableValueLabel = ({
+  allowEdit,
+  isNew,
+  onCancel,
+  onEdit,
+  onSave,
+  validationFunc,
+  value,
+}: IEditableValueLabel) => {
+  const valueElement = useRef<HTMLSpanElement>(null);
+  const inputElement = useRef<HTMLInputElement>(null);
+
+  const [editing, setEditing] = useState(isNew);
+  const [
+    { internalValue, isValid, validationError },
+    setInternalValue,
+  ] = useReducer<Reducer<IReducerState, string>>(
```

Comment:
I didn't know about this hook yet :P 
took me a while to get it.

Maybe for understandability, could we extract the arguments into named variables like `reducer` and `initialValue` or something like that?

I don't think prettier makes it easier to read this stuff either.. what do you guys think? Is it something I just have to get used to?

Reply:
This is indeed hard to read. I can try to move this in separate functions.